src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs

```diff
@@ -1429,15 +1429,36 @@ public static class PipelineStageDefinitionBuilder
         }
 
         // private methods
-        private static bool IsTConnectToOrEnumerableTConnectTo<TConnectFrom, TConnectTo>()
+        private static bool AreGraphLookupTypesCompatible<TConnectFrom, TConnectTo>()
```

Comment:
The agreed upon name was `AreGroupLookupFromAndToTypesCompatible`.

Reply:
I couldn't convince myself that the changes here were valid. I was expecting some code that looked more like the English description of the constraints, i.e. `AreGroupLookupFromAndToTypesCompatible` should return true when one of the following is true

1. connectFromType == connectToType
2. connectFromType and connectToType are BOTH IEnumerable and have the same item type
3. only connectFromType is IEnumerable and fromItemType == connectToType
4. only connectToType is IEnumerable and connectFromType == toItemType
5. otherwise return false

Reply:
Expected code something like:

    private static bool AreGroupLookupFromAndToTypesCompatible<TConnectFrom, TConnectTo>(string fromGenericTypeParamName)
    {
        if (typeof(TConnectFrom) == typeof(TConnectTo))
        {
            return true;
        }

        var fromIEnumerableType = FindIEnumerableType(typeof(TConnectFrom), fromGenericTypeParamName);
        var toIEnumerableType = FindIEnumerableType(typeof(TConnectTo), "TConnectTo");

        if (fromIEnumerableType != null && toIEnumerableType != null)
        {
            var fromItemType = GetItemType(fromIEnumerableType);
            var toItemType = GetItemType(toIEnumerableType);
            return fromItemType == toItemType;
        }
        else if (fromIEnumerableType != null)
        {
            var fromItemType = GetItemType(fromIEnumerableType);
            return fromItemType == typeof(TConnectTo);
        }
        else if (toIEnumerableType != null)
        {
            var toItemType = GetItemType(toIEnumerableType);
            return typeof(TConnectFrom) == toItemType;
        }
        else
        {
            return false;
        }

        Type FindIEnumerableType(Type type, string genericTypeParameterName)
        {
            var iEnumerablenterfaces = type.GetTypeInfo()
                .GetInterfaces()
                .Where(i => i.Name == "IEnumerable" && i.IsConstructedGenericType && i.GenericTypeArguments.Length == 1)
                .ToList();
            switch (iEnumerablenterfaces.Count)
            {
                case 0:
                    return null;
                case 1:
                    return iEnumerablenterfaces.Single();
                default:
                    var message = $"{genericTypeParameterName} cannot implement more than one IEnumerable<T> interface.";
                    throw new ArgumentException(message, genericTypeParameterName);
            }
        }

        Type GetItemType(Type ienumerableType)
        {
            return ienumerableType.GenericTypeArguments.Single();
        }
    }

Reply:
Thinking about it some more I'm not sure that code is right either.

I think we need to think harder about the English description of the type constraints.

What EXACTLY makes TConnectFrom and TConnectTo compatible?

Reply:
Here's another attempt at `AreGroupLookupFromAndToTypesCompatible` using `IsAssignableFrom` to allow GraphLookup to work with polymorphic from and to types (for example `Manager` and `Employee`).

    private static bool AreGroupLookupFromAndToTypesCompatible<TConnectFrom, TConnectTo>(string fromGenericTypeParamName)
    {
        var connectFromType = typeof(TConnectFrom);
        var connectToType = typeof(TConnectTo);

        if (IsAssignableFrom(connectFromType, connectToType) || IsAssignableFrom(connectToType, connectFromType))
        {
            return true;
        }

        var enumerableConnectFromType = typeof(IEnumerable<>).MakeGenericType(connectFromType);
        if (IsAssignableFrom(enumerableConnectFromType, connectToType))
        {
            return true;
        }

        var enumerableConnectToType = typeof(IEnumerable<>).MakeGenericType(connectToType);
        if (IsAssignableFrom(enumerableConnectToType, connectFromType))
        {
            return true;
        }

        var connectFromItemType = GetItemTypeIfEnumerable(connectFromType, fromGenericTypeParamName);
        var connectToItemType = GetItemTypeIfEnumerable(connectToType, "TConnectTo");
        if (connectFromItemType != null && connectToItemType != null)
        {
            if (IsAssignableFrom(connectFromItemType, connectToItemType) || IsAssignableFrom(connectToItemType, connectFromItemType))
            {
                return true;
            }
        }

        return false;

        bool IsAssignableFrom(Type type1, Type type2)
        {
            return type1.GetTypeInfo().IsAssignableFrom(type2);
        }

        Type GetItemTypeIfEnumerable(Type type, string genericTypeParameterName)
        {
            var iEnumerablenterfaces = type.GetTypeInfo()
                .GetInterfaces()
                .Where(i => i.Name == "IEnumerable" && i.IsConstructedGenericType && i.GenericTypeArguments.Length == 1)
                .ToList();
            switch (iEnumerablenterfaces.Count)
            {
                case 0:
                    return null;
                case 1:
                    var ienumerableType = iEnumerablenterfaces.Single();
                    return ienumerableType.GenericTypeArguments.Single();
                default:
                    var message = $"{genericTypeParameterName} cannot implement more than one IEnumerable<T> interface.";
                    throw new ArgumentException(message, genericTypeParameterName);
            }
        }
    }

Reply:
As mentioned in an earlier comment, we need to think really hard about EXACTLY what makes from and to types compatible...

Reply:
Resolving this comment because we aren't going to handle many to many any more.